src/java.base/share/classes/jdk/internal/constant/ReferenceClassDescImpl.java

```diff
@@ -50,11 +48,14 @@ private ReferenceClassDescImpl(String descriptor) {
      * field descriptor string, or does not describe a class or interface type
      * @jvms 4.3.2 Field Descriptors
      */
-    public static ReferenceClassDescImpl of(String descriptor) {
+    public static ClassDesc of(String descriptor) {
         int dLen = descriptor.length();
         int len = ConstantUtils.skipOverFieldSignature(descriptor, 0, dLen);
         if (len <= 1 || len != dLen)
             throw new IllegalArgumentException(String.format("not a valid reference type descriptor: %s", descriptor));
+        if (descriptor.charAt(0) == '[') {
+            return ArrayClassDescImpl.ofValidatedDescriptor(descriptor);
+        }
```

Comment:
Indeed, since the factory usages are widespread, I only did a hotfix migration. This class should now be called `ClassOrInterfaceDescImpl` (or "type klass" in hotspot terms or "declared types" in javac terms)